File: tythe_time_tracker/database/repository.py
"""Repository pattern for time entry data access operations."""

import logging
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from psycopg2.extras import Json

from ..core.constants import DatabaseConstants, PayRateType
from ..core.models import TimeEntry
from .connection import DatabaseConnection

logger = logging.getLogger(__name__)


class TimeEntryRepository:
    """Repository for time entry data access operations."""
    
    def __init__(self, db_connection: DatabaseConnection) -> None:
        """Initialize the repository with a database connection.
        
        Args:
            db_connection: Database connection instance.
        """
        self.db = db_connection
    
    def create_time_entry(
        self, 
        employee: str, 
        clock_in: datetime, 
        pay_rate_type: PayRateType,
        clock_out: Optional[datetime] = None
    ) -> TimeEntry:
        """Create a new time entry.
        
        Args:
            employee: Employee name.
            clock_in: Clock in time.
            pay_rate_type: Pay rate type for the entry.
            clock_out: Optional clock out time.
            
        Returns:
            The created TimeEntry instance.
            
        Raises:
            ValueError: If employee name is empty.
        """
        if not employee.strip():
            raise ValueError("Employee name cannot be empty")
        
        try:
            with self.db.get_cursor() as cursor:
                cursor.execute(f"""
                    INSERT INTO {DatabaseConstants.TIME_ENTRIES_TABLE} 
                    ({DatabaseConstants.EMPLOYEE_COLUMN}, {DatabaseConstants.CLOCK_IN_COLUMN}, 
                     {DatabaseConstants.CLOCK_OUT_COLUMN}, {DatabaseConstants.PAY_RATE_TYPE_COLUMN})
                    VALUES (%s, %s, %s, %s)
                    RETURNING {DatabaseConstants.ID_COLUMN}, {DatabaseConstants.CREATED_AT_COLUMN}
                """, (employee.strip(), clock_in, clock_out, pay_rate_type.value))
                
                entry_id, created_at = cursor.fetchone()
                
                return TimeEntry(
                    id=entry_id,
                    employee=employee.strip(),
                    clock_in=clock_in,
                    clock_out=clock_out,
                    pay_rate_type=pay_rate_type,
                    created_at=created_at
                )
        except Exception as e:
            logger.error(f"Failed to create time entry: {e}")
            raise
    
    def get_open_shift(self, employee: str) -> Optional[TimeEntry]:
        """Get the most recent open shift for an employee.
        
        Args:
            employee: Employee name.
            
        Returns:
            The open TimeEntry if found, None otherwise.
        """
        try:
            with self.db.get_cursor() as cursor:
                cursor.execute(f"""
                    SELECT {DatabaseConstants.ID_COLUMN}, {DatabaseConstants.EMPLOYEE_COLUMN},
                           {DatabaseConstants.CLOCK_IN_COLUMN}, {DatabaseConstants.CLOCK_OUT_COLUMN},
                           {DatabaseConstants.PAY_RATE_TYPE_COLUMN}, {DatabaseConstants.CREATED_AT_COLUMN}
                    FROM {DatabaseConstants.TIME_ENTRIES_TABLE}
                    WHERE LOWER({DatabaseConstants.EMPLOYEE_COLUMN}) = LOWER(%s) 
                    AND {DatabaseConstants.CLOCK_OUT_COLUMN} IS NULL
                    ORDER BY {DatabaseConstants.CLOCK_IN_COLUMN} DESC
                    LIMIT 1
                """, (employee.strip(),))
                
                row = cursor.fetchone()
                if row:
                    return self._row_to_time_entry(row)
                return None
        except Exception as e:
            logger.error(f"Failed to get open shift: {e}")
            raise
    
    def close_shift(self, entry_id: UUID, clock_out: datetime) -> TimeEntry:
        """Close a time entry by setting the clock out time.
        
        Args:
            entry_id: The ID of the time entry to close.
            clock_out: The clock out time.
            
        Returns:
            The updated TimeEntry instance.
            
        Raises:
            ValueError: If the time entry is not found.
        """
        try:
            with self.db.get_cursor() as cursor:
                cursor.execute(f"""
                    UPDATE {DatabaseConstants.TIME_ENTRIES_TABLE}
                    SET {DatabaseConstants.CLOCK_OUT_COLUMN} = %s
                    WHERE {DatabaseConstants.ID_COLUMN} = %s
                    RETURNING {DatabaseConstants.ID_COLUMN}, {DatabaseConstants.EMPLOYEE_COLUMN},
                             {DatabaseConstants.CLOCK_IN_COLUMN}, {DatabaseConstants.CLOCK_OUT_COLUMN},
                             {DatabaseConstants.PAY_RATE_TYPE_COLUMN}, {DatabaseConstants.CREATED_AT_COLUMN}
                """, (clock_out, entry_id))
                
                row = cursor.fetchone()
                if not row:
                    raise ValueError(f"Time entry with ID {entry_id} not found")
                
                return self._row_to_time_entry(row)
        except Exception as e:
            logger.error(f"Failed to close shift: {e}")
            raise
    
    def get_employee_timesheet(
        self, 
        employee: str, 
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None
    ) -> List[TimeEntry]:
        """Get timesheet entries for a specific employee.
        
        Args:
            employee: Employee name.
            start_date: Optional start date filter.
            end_date: Optional end date filter.
            
        Returns:
            List of TimeEntry instances.
        """
        try:
            with self.db.get_cursor() as cursor:
                query = f"""
                    SELECT {DatabaseConstants.ID_COLUMN}, {DatabaseConstants.EMPLOYEE_COLUMN},
                           {DatabaseConstants.CLOCK_IN_COLUMN}, {DatabaseConstants.CLOCK_OUT_COLUMN},
                           {DatabaseConstants.PAY_RATE_TYPE_COLUMN}, {DatabaseConstants.CREATED_AT_COLUMN}
                    FROM {DatabaseConstants.TIME_ENTRIES_TABLE}
                    WHERE LOWER({DatabaseConstants.EMPLOYEE_COLUMN}) = LOWER(%s)
                """
                params = [employee.strip()]
                
                if start_date:
                    query += f" AND DATE({DatabaseConstants.CLOCK_IN_COLUMN}) >= %s"
                    params.append(start_date.date())
                
                if end_date:
                    query += f" AND DATE({DatabaseConstants.CLOCK_IN_COLUMN}) <= %s"
                    params.append(end_date.date())
                
                query += f" ORDER BY {DatabaseConstants.CLOCK_IN_COLUMN} DESC"
                
                cursor.execute(query, params)
                rows = cursor.fetchall()
                
                return [self._row_to_time_entry(row) for row in rows]
        except Exception as e:
            logger.error(f"Failed to get employee timesheet: {e}")
            raise
    
    def get_all_timesheets(
        self,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None
    ) -> List[TimeEntry]:
        """Get all timesheet entries with optional date filters.
        
        Args:
            start_date: Optional start date filter.
            end_date: Optional end date filter.
            
        Returns:
            List of TimeEntry instances.
        """
        try:
            with self.db.get_cursor() as cursor:
                query = f"""
                    SELECT {DatabaseConstants.ID_COLUMN}, {DatabaseConstants.EMPLOYEE_COLUMN},
                           {DatabaseConstants.CLOCK_IN_COLUMN}, {DatabaseConstants.CLOCK_OUT_COLUMN},
                           {DatabaseConstants.PAY_RATE_TYPE_COLUMN}, {DatabaseConstants.CREATED_AT_COLUMN}
                    FROM {DatabaseConstants.TIME_ENTRIES_TABLE}
                    WHERE 1=1
                """
                params = []
                
                if start_date:
                    query += f" AND DATE({DatabaseConstants.CLOCK_IN_COLUMN}) >= %s"
                    params.append(start_date.date())
                
                if end_date:
                    query += f" AND DATE({DatabaseConstants.CLOCK_IN_COLUMN}) <= %s"
                    params.append(end_date.date())
                
                query += f" ORDER BY {DatabaseConstants.CLOCK_IN_COLUMN} DESC"
                
                cursor.execute(query, params)
                rows = cursor.fetchall()
                
                return [self._row_to_time_entry(row) for row in rows]
        except Exception as e:
            logger.error(f"Failed to get all timesheets: {e}")
            raise
    
    def get_time_entry_by_id(self, entry_id: UUID) -> Optional[TimeEntry]:
        """Get a time entry by its ID.
        
        Args:
            entry_id: The ID of the time entry.
            
        Returns:
            The TimeEntry if found, None otherwise.
        """
        try:
            with self.db.get_cursor() as cursor:
                cursor.execute(f"""
                    SELECT {DatabaseConstants.ID_COLUMN}, {DatabaseConstants.EMPLOYEE_COLUMN},
                           {DatabaseConstants.CLOCK_IN_COLUMN}, {DatabaseConstants.CLOCK_OUT_COLUMN},
                           {DatabaseConstants.PAY_RATE_TYPE_COLUMN}, {DatabaseConstants.CREATED_AT_COLUMN}
                    FROM {DatabaseConstants.TIME_ENTRIES_TABLE}
                    WHERE {DatabaseConstants.ID_COLUMN} = %s
                """, (entry_id,))
                
                row = cursor.fetchone()
                if row:
                    return self._row_to_time_entry(row)
                return None
        except Exception as e:
            logger.error(f"Failed to get time entry by ID: {e}")
            raise
    
    def update_time_entry(
        self,
        entry_id: UUID,
        employee: str,
        clock_in: datetime,
        clock_out: Optional[datetime],
        pay_rate_type: PayRateType
    ) -> TimeEntry:
        """Update an existing time entry.
        
        Args:
            entry_id: The ID of the time entry to update.
            employee: Updated employee name.
            clock_in: Updated clock in time.
            clock_out: Updated clock out time.
            pay_rate_type: Updated pay rate type.
            
        Returns:
            The updated TimeEntry instance.
            
        Raises:
            ValueError: If the time entry is not found.
        """
        try:
            with self.db.get_cursor() as cursor:
                cursor.execute(f"""
                    UPDATE {DatabaseConstants.TIME_ENTRIES_TABLE}
                    SET {DatabaseConstants.EMPLOYEE_COLUMN} = %s,
                        {DatabaseConstants.CLOCK_IN_COLUMN} = %s,
                        {DatabaseConstants.CLOCK_OUT_COLUMN} = %s,
                        {DatabaseConstants.PAY_RATE_TYPE_COLUMN} = %s
                    WHERE {DatabaseConstants.ID_COLUMN} = %s
                    RETURNING {DatabaseConstants.ID_COLUMN}, {DatabaseConstants.EMPLOYEE_COLUMN},
                             {DatabaseConstants.CLOCK_IN_COLUMN}, {DatabaseConstants.CLOCK_OUT_COLUMN},
                             {DatabaseConstants.PAY_RATE_TYPE_COLUMN}, {DatabaseConstants.CREATED_AT_COLUMN}
                """, (employee.strip(), clock_in, clock_out, pay_rate_type.value, entry_id))
                
                row = cursor.fetchone()
                if not row:
                    raise ValueError(f"Time entry with ID {entry_id} not found")
                
                return self._row_to_time_entry(row)
        except Exception as e:
            logger.error(f"Failed to update time entry: {e}")
            raise
    
    def delete_time_entry(self, entry_id: UUID) -> bool:
        """Delete a time entry.
        
        Args:
            entry_id: The ID of the time entry to delete.
            
        Returns:
            True if the entry was deleted, False if not found.
        """
        try:
            with self.db.get_cursor() as cursor:
                cursor.execute(f"""
                    DELETE FROM {DatabaseConstants.TIME_ENTRIES_TABLE}
                    WHERE {DatabaseConstants.ID_COLUMN} = %s
                """, (entry_id,))
                
                return cursor.rowcount > 0
        except Exception as e:
            logger.error(f"Failed to delete time entry: {e}")
            raise

    def insert_audit_log(
        self,
        action: str,
        target_table: str,
        target_id: Optional[UUID],
        changed_by: str,
        old_values: Optional[Dict[str, Any]] = None,
        new_values: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Insert an audit log record and return key fields."""
        try:
            with self.db.get_cursor() as cursor:
                cursor.execute(
                    f"""
                    INSERT INTO {DatabaseConstants.AUDIT_LOG_TABLE}
                        (action, target_table, target_id, changed_by, old_values, new_values)
                    VALUES (%s, %s, %s, %s, %s, %s)
                    RETURNING id, created_at
                    """,
                    (
                        action,
                        target_table,
                        target_id,
                        changed_by.strip(),
                        Json(old_values) if old_values is not None else None,
                        Json(new_values) if new_values is not None else None,
                    ),
                )
                row = cursor.fetchone()
                if row is None:
                    raise RuntimeError("Failed to insert audit log")
                audit_id, created_at = row
                return {"id": audit_id, "created_at": created_at}
        except Exception as e:
            logger.error(f"Failed to insert audit log: {e}")
            raise

    def get_audit_logs(self, filters: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """Query audit log records with optional filters.

        Supported filters: action, target_table, target_id, changed_by,
        start_date, end_date, limit.
        """
        filters = filters or {}
        try:
            with self.db.get_cursor() as cursor:
                query = f"""
                    SELECT id, action, target_table, target_id, changed_by,
                           old_values, new_values, created_at
                    FROM {DatabaseConstants.AUDIT_LOG_TABLE}
                    WHERE 1=1
                """
                params: List[Any] = []

                if filters.get("action"):
                    query += " AND action = %s"
                    params.append(filters["action"])
                if filters.get("target_table"):
                    query += " AND target_table = %s"
                    params.append(filters["target_table"])
                if filters.get("target_id"):
                    query += " AND target_id = %s"
                    params.append(filters["target_id"])
                if filters.get("changed_by"):
                    query += " AND changed_by = %s"
                    params.append(filters["changed_by"])
                if filters.get("start_date"):
                    query += " AND created_at >= %s"
                    params.append(filters["start_date"])
                if filters.get("end_date"):
                    query += " AND created_at <= %s"
                    params.append(filters["end_date"])

                query += " ORDER BY created_at DESC"
                if filters.get("limit"):
                    query += " LIMIT %s"
                    params.append(int(filters["limit"]))

                cursor.execute(query, params)
                rows = cursor.fetchall()

                return [
                    {
                        "id": row[0],
                        "action": row[1],
                        "target_table": row[2],
                        "target_id": row[3],
                        "changed_by": row[4],
                        "old_values": row[5],
                        "new_values": row[6],
                        "created_at": row[7],
                    }
                    for row in rows
                ]
        except Exception as e:
            logger.error(f"Failed to get audit logs: {e}")
            raise
    
    def _row_to_time_entry(self, row: tuple) -> TimeEntry:
        """Convert a database row to a TimeEntry instance.
        
        Args:
            row: Database row tuple.
            
        Returns:
            TimeEntry instance.
        """
        entry_id, employee, clock_in, clock_out, pay_rate_type, created_at = row
        return TimeEntry(
            id=entry_id,
            employee=employee,
            clock_in=clock_in,
            clock_out=clock_out,
            pay_rate_type=PayRateType(pay_rate_type),
            created_at=created_at
        ) 

File: tt-ts/src/server/db/repository.ts
import type { TimeEntry, PayRateType, AuditLogEntry } from '../../shared/types.js'
import { DB } from '../../shared/constants.js'

type TimeEntryRow = {
  id: string
  user_id: string | null
  employee: string
  clock_in: Date
  clock_out: Date | null
  pay_rate_type: string
  created_at: Date
}

function rowToTimeEntry(row: {
  id: string
  user_id: string | null
  employee: string
  clock_in: Date
  clock_out: Date | null
  pay_rate_type: string
  created_at: Date
}): TimeEntry {
  return {
    id: row.id,
    user_id: row.user_id ?? null,
    employee: row.employee,
    clock_in: new Date(row.clock_in),
    clock_out: row.clock_out ? new Date(row.clock_out) : null,
    pay_rate_type: row.pay_rate_type as PayRateType,
    created_at: new Date(row.created_at),
  }
}

export async function createTimeEntry(
  employee: string,
  clockIn: Date,
  payRateType: PayRateType,
  clockOut?: Date | null,
  userId?: string | null
): Promise<TimeEntry> {
  const { query } = await import('./connection.js')
  const res = await query<TimeEntryRow>(
    `INSERT INTO ${DB.TIME_ENTRIES_TABLE} 
     (${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN})
     VALUES ($1, $2, $3, $4, $5)
     RETURNING ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}`,
    [userId ?? null, employee.trim(), clockIn, clockOut ?? null, payRateType]
  )
  return rowToTimeEntry(res.rows[0])
}

export async function getOpenShift(employee: string): Promise<TimeEntry | null> {
  const { query } = await import('./connection.js')
  const res = await query<TimeEntryRow>(
    `SELECT ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}
     FROM ${DB.TIME_ENTRIES_TABLE}
     WHERE LOWER(${DB.EMPLOYEE_COLUMN}) = LOWER($1) AND ${DB.CLOCK_OUT_COLUMN} IS NULL
     ORDER BY ${DB.CLOCK_IN_COLUMN} DESC LIMIT 1`,
    [employee.trim()]
  )
  return res.rows[0] ? rowToTimeEntry(res.rows[0]) : null
}

export async function getOpenShiftByUserId(
  userId: string,
  fallbackEmployee?: string | null
): Promise<TimeEntry | null> {
  const { query } = await import('./connection.js')
  const params: unknown[] = [userId]
  let sql = `SELECT ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}
     FROM ${DB.TIME_ENTRIES_TABLE}
     WHERE ${DB.USER_ID_COLUMN} = $1 AND ${DB.CLOCK_OUT_COLUMN} IS NULL`
  if (fallbackEmployee?.trim()) {
    params.push(fallbackEmployee.trim())
    sql += ` OR (${DB.USER_ID_COLUMN} IS NULL AND LOWER(${DB.EMPLOYEE_COLUMN}) = LOWER($${params.length}) AND ${DB.CLOCK_OUT_COLUMN} IS NULL)`
  }
  sql += ` ORDER BY ${DB.CLOCK_IN_COLUMN} DESC LIMIT 1`
  const res = await query<TimeEntryRow>(sql, params)
  return res.rows[0] ? rowToTimeEntry(res.rows[0]) : null
}

export async function closeShift(entryId: string, clockOut: Date): Promise<TimeEntry> {
  const { query } = await import('./connection.js')
  const res = await query<TimeEntryRow>(
    `UPDATE ${DB.TIME_ENTRIES_TABLE} SET ${DB.CLOCK_OUT_COLUMN} = $1 WHERE ${DB.ID_COLUMN} = $2
     RETURNING ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}`,
    [clockOut, entryId]
  )
  if (!res.rows[0]) throw new Error(`Time entry ${entryId} not found`)
  return rowToTimeEntry(res.rows[0])
}

export async function getEmployeeTimesheet(
  employee: string,
  startDate?: Date | null,
  endDate?: Date | null
): Promise<TimeEntry[]> {
  const { query } = await import('./connection.js')
  let sql = `SELECT ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}
     FROM ${DB.TIME_ENTRIES_TABLE} WHERE LOWER(${DB.EMPLOYEE_COLUMN}) = LOWER($1)`
  const params: unknown[] = [employee.trim()]
  if (startDate) {
    params.push(startDate)
    sql += ` AND DATE(${DB.CLOCK_IN_COLUMN}) >= $${params.length}`
  }
  if (endDate) {
    params.push(endDate)
    sql += ` AND DATE(${DB.CLOCK_IN_COLUMN}) <= $${params.length}`
  }
  sql += ` ORDER BY ${DB.CLOCK_IN_COLUMN} DESC`
  const res = await query<TimeEntryRow>(sql, params)
  return res.rows.map(rowToTimeEntry)
}

export async function getEmployeeTimesheetByUserId(
  userId: string,
  fallbackEmployee?: string | null,
  startDate?: Date | null,
  endDate?: Date | null
): Promise<TimeEntry[]> {
  const { query } = await import('./connection.js')
  const params: unknown[] = [userId]
  let sql = `SELECT ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}
     FROM ${DB.TIME_ENTRIES_TABLE}
     WHERE (${DB.USER_ID_COLUMN} = $1`
  if (fallbackEmployee?.trim()) {
    params.push(fallbackEmployee.trim())
    sql += ` OR (${DB.USER_ID_COLUMN} IS NULL AND LOWER(${DB.EMPLOYEE_COLUMN}) = LOWER($${params.length}))`
  }
  sql += `)`
  if (startDate) {
    params.push(startDate)
    sql += ` AND DATE(${DB.CLOCK_IN_COLUMN}) >= $${params.length}`
  }
  if (endDate) {
    params.push(endDate)
    sql += ` AND DATE(${DB.CLOCK_IN_COLUMN}) <= $${params.length}`
  }
  sql += ` ORDER BY ${DB.CLOCK_IN_COLUMN} DESC`
  const res = await query<TimeEntryRow>(sql, params)
  return res.rows.map(rowToTimeEntry)
}

export async function getAllTimesheets(
  startDate?: Date | null,
  endDate?: Date | null
): Promise<TimeEntry[]> {
  const { query } = await import('./connection.js')
  let sql = `SELECT ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}
     FROM ${DB.TIME_ENTRIES_TABLE} WHERE 1=1`
  const params: unknown[] = []
  if (startDate) {
    params.push(startDate)
    sql += ` AND DATE(${DB.CLOCK_IN_COLUMN}) >= $${params.length}`
  }
  if (endDate) {
    params.push(endDate)
    sql += ` AND DATE(${DB.CLOCK_IN_COLUMN}) <= $${params.length}`
  }
  sql += ` ORDER BY ${DB.CLOCK_IN_COLUMN} DESC`
  const res = await query<TimeEntryRow>(sql, params)
  return res.rows.map(rowToTimeEntry)
}

export async function getTimeEntryById(entryId: string): Promise<TimeEntry | null> {
  const { query } = await import('./connection.js')
  const res = await query<TimeEntryRow>(
    `SELECT ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}
     FROM ${DB.TIME_ENTRIES_TABLE} WHERE ${DB.ID_COLUMN} = $1`,
    [entryId]
  )
  return res.rows[0] ? rowToTimeEntry(res.rows[0]) : null
}

export async function updateTimeEntry(
  entryId: string,
  employee: string,
  clockIn: Date,
  clockOut: Date | null,
  payRateType: PayRateType
): Promise<TimeEntry> {
  const { query } = await import('./connection.js')
  const res = await query<TimeEntryRow>(
    `UPDATE ${DB.TIME_ENTRIES_TABLE}
     SET ${DB.EMPLOYEE_COLUMN} = $1, ${DB.CLOCK_IN_COLUMN} = $2, ${DB.CLOCK_OUT_COLUMN} = $3, ${DB.PAY_RATE_TYPE_COLUMN} = $4
     WHERE ${DB.ID_COLUMN} = $5
     RETURNING ${DB.ID_COLUMN}, ${DB.USER_ID_COLUMN}, ${DB.EMPLOYEE_COLUMN}, ${DB.CLOCK_IN_COLUMN}, ${DB.CLOCK_OUT_COLUMN}, ${DB.PAY_RATE_TYPE_COLUMN}, ${DB.CREATED_AT_COLUMN}`,
    [employee.trim(), clockIn, clockOut, payRateType, entryId]
  )
  if (!res.rows[0]) throw new Error(`Time entry ${entryId} not found`)
  return rowToTimeEntry(res.rows[0])
}

export async function deleteTimeEntry(entryId: string): Promise<boolean> {
  const { query } = await import('./connection.js')
  const res = await query(
    `DELETE FROM ${DB.TIME_ENTRIES_TABLE} WHERE ${DB.ID_COLUMN} = $1`,
    [entryId]
  )
  return (res.rowCount ?? 0) > 0
}

export async function insertAuditLog(
  action: 'add' | 'edit' | 'delete',
  targetTable: string,
  targetId: string | null,
  changedBy: string,
  oldValues?: Record<string, unknown> | null,
  newValues?: Record<string, unknown> | null
): Promise<{ id: string; created_at: Date }> {
  const { query } = await import('./connection.js')
  const res = await query<{ id: string; created_at: Date }>(
    `INSERT INTO ${DB.AUDIT_LOG_TABLE} (action, target_table, target_id, changed_by, old_values, new_values)
     VALUES ($1, $2, $3, $4, $5, $6)
     RETURNING id, created_at`,
    [action, targetTable, targetId, changedBy.trim(), oldValues ? JSON.stringify(oldValues) : null, newValues ? JSON.stringify(newValues) : null]
  )
  return res.rows[0]
}

export async function getAuditLogs(filters?: {
  action?: string
  target_table?: string
  changed_by?: string
  start_date?: Date
  end_date?: Date
  limit?: number
}): Promise<AuditLogEntry[]> {
  const { query } = await import('./connection.js')
  let sql = `SELECT id, action, target_table, target_id, changed_by, old_values, new_values, created_at
     FROM ${DB.AUDIT_LOG_TABLE} WHERE 1=1`
  const params: unknown[] = []
  if (filters?.action) {
    params.push(filters.action)
    sql += ` AND action = $${params.length}`
  }
  if (filters?.target_table) {
    params.push(filters.target_table)
    sql += ` AND target_table = $${params.length}`
  }
  if (filters?.changed_by) {
    params.push(filters.changed_by)
    sql += ` AND changed_by = $${params.length}`
  }
  if (filters?.start_date) {
    params.push(filters.start_date)
    sql += ` AND created_at >= $${params.length}`
  }
  if (filters?.end_date) {
    params.push(filters.end_date)
    sql += ` AND created_at <= $${params.length}`
  }
  sql += ' ORDER BY created_at DESC'
  if (filters?.limit) {
    params.push(filters.limit)
    sql += ` LIMIT $${params.length}`
  }
  const res = await query<{
    id: string
    action: 'add' | 'edit' | 'delete'
    target_table: string
    target_id: string | null
    changed_by: string
    old_values: unknown
    new_values: unknown
    created_at: Date
  }>(sql, params)
  return res.rows.map((r) => ({
    id: r.id,
    action: r.action,
    target_table: r.target_table,
    target_id: r.target_id,
    changed_by: r.changed_by,
    old_values: r.old_values as Record<string, unknown> | null,
    new_values: r.new_values as Record<string, unknown> | null,
    created_at: new Date(r.created_at),
  }))
}
